### PR TITLE
[WIP] add minimum prompt editor

### PR DESCRIPTION
**Pull request by thesophiaxu**
Putting it here mainly so that y'all can see a preview for the editor.

To trigger it, currently you have to hold Cmd/Ctrl and click on text in the right sidebar (doens't work if you're clicking on spaces)

### CODE REVIEW

File: ui/components/TracePage/TracePage.tsx [hunk 1]
```diff
@@ -832,14 +842,17 @@ export const TracePage = () => {
...
-        </title>
-      </Head>
-      <Trace traceId={traceId} />
+      <PromptEditorProvider>
+        <Head>
+          <title>
+            {traceId && recipes[traceId]
+              ? `${recipes[traceId].title} | Interactive Composition Explorer`
+              : "Interactive Composition Explorer"}
+          </title>
```
Comment (lslunis) [inline ui/components/TracePage/TracePage.tsx:851]: Might be worth extracting <title> into a component

File: ui/components/TracePage/PromptEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+const PromptEditorContext = createContext<{
+  open: boolean;
+  currentPrompt: string;
+  setCurrentPrompt: (val: string | ((prev: string) => string)) => void;
+  openEditor: (content: string) => void;
+  closeEditor: () => void;
+} | null>(null);
+
+function saveCaretPosition(context: any, plusOne?: boolean) {
+  var sel = window.getSelection();
```
Comment (lslunis) [inline ui/components/TracePage/PromptEditor.tsx:21]: These should be `const` if they aren't reassigned and `let` if they are.

File: ui/.env [hunk 3]
```diff
@@ -0,0 +1 @@
+NEXT_PUBLIC_BACKEND_URL=https://thesophiaxu-oughtinc-ice-6vxpvq6prr52x4wq-8935.githubpreview.dev
```
Comment (lslunis) [inline ui/.env:1]: In `.gitignore`, can you change `/.env` to `**/.env`, and then revert this file?

Comment (thesophiaxu) [reply ui/.env]: Done

File: ui/components/TracePage/PromptEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+              }}
+            >
+              {loading ? "Loading..." : "Prompt model"}
+            </Button>
+          </div>
+        </div>
+      </div>
+    </div>
+  );
+};
```
Comment (lslunis) [inline ui/components/TracePage/PromptEditor.tsx:236]: This component is almost 150 lines long. Do you think there are some functions, hooks, or subcomponents that could be cleanly factored out?

File: ui/components/TracePage/PromptEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+                    "Content-Type": "application/json",
+                  },
+                  body: JSON.stringify({
+                    agent: selectedAgent,
+                    prompt: currentPrompt,
+                    multiline,
+                  }),
+                })
+                  .then(res => res.json())
+                  .then(data => {
```
Comment (lslunis) [inline ui/components/TracePage/PromptEditor.tsx:223]: Let's use `await` instead of `.then` (change all)

**Review by lslunis — COMMENTED** [review]
Backend looks good!

I'll take a close look at the frontend once the comments are addressed, but the general approach seems good

**Review by lslunis — COMMENTED** [review]
Could you add some tests?

File: ui/components/TracePage/PromptEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+              className="rounded-4 px-2 py-1 h-fit !shadow-none hover:bg-slate-200 mt-4 text-xs"
+              size="md"
+              variant="outline"
+              onClick={() => {
+                setLoading(true);
+                fetch(`${backendUrl}/agents/complete`, {
+                  method: "POST",
+                  headers: {
+                    "Content-Type": "application/json",
+                    "X-Api-Key": process.env.NEXT_PUBLIC_API_KEY ?? "",
```
Comment (lslunis) [inline ui/components/TracePage/PromptEditor.tsx:215]:
```suggestion
                    "X-Api-Key": process.env. NEXT_PUBLIC_BACKEND_URL ?? "",
```

File: ui/components/TracePage/PromptEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+              className="rounded-4 px-2 py-1 h-fit !shadow-none hover:bg-slate-200 mt-4 text-xs"
+              size="md"
+              variant="outline"
+              onClick={() => {
+                setLoading(true);
+                fetch(`${backendUrl}/agents/complete`, {
+                  method: "POST",
+                  headers: {
+                    "Content-Type": "application/json",
+                    "X-Api-Key": process.env.NEXT_PUBLIC_BACKEND_URL ?? "",
```
Comment (lslunis) [inline ui/components/TracePage/PromptEditor.tsx:215]:
```suggestion
                    "X-Api-Key": process.env.NEXT_PUBLIC_BACKEND_API_KEY ?? "",
```

File: ice/routes/agents.py [hunk 8]
```diff
@@ -1,19 +1,28 @@
...
-async def list():
-    return ["alice", "bob"]
+async def apiList():
+    return list(MACHINE_AGENTS.keys())
 
 
 class CompleteRequest(RouteModel):
+    agent: str
     prompt: str
+    multiline: bool
```
[the remark this answers is not in the document]
Comment (stuhlmueller) [reply ice/routes/agents.py:19]: Yes, use `stop` value like in the agent interface now.

File: ui/components/TracePage/TracePage.tsx (hunk 1 above)
Comment (thesophiaxu) [reply ui/components/TracePage/TracePage.tsx:851]: We can probably do that at a later date, but it's out of scope with this PR

File: ui/components/TracePage/PromptEditor.tsx (hunk 5 above)
Comment (thesophiaxu) [reply ui/components/TracePage/PromptEditor.tsx]: done

File: ui/components/TracePage/PromptEditor.tsx (hunk 4 above)
Comment (thesophiaxu) [reply ui/components/TracePage/PromptEditor.tsx:379]: put the contenteditable out - that shouldn't be entangled but other stuff can be quite connected

File: ice/routes/agents.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (thesophiaxu) [reply ice/routes/agents.py]: Resolved in the latest version

File: ice/routes/agents.py [hunk 9]
```diff
@@ -1,19 +1,28 @@
...
 
 
 @router.post("/complete")
 async def complete(request: CompleteRequest):
-    return "hi"
+    if request.agent not in MACHINE_AGENTS:
+        return "Invalid agent!"
```
[the remark this answers is not in the document]
Comment (thesophiaxu) [reply ice/routes/agents.py]: done

**Review by lslunis — COMMENTED** [review]
Looks good! I still have to review PromptEditor.tsx, doing that now.

File: ui/components/TracePage/PromptEditor.tsx [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+          <label className="text-sm mb-1 mt-3 text-gray-700">Task</label>
+          <Select onChange={ev => setTask(ev.target.value)} value={task}>
+            <option value="complete">Complete</option>
+            <option value="classify">Classify</option>
+          </Select>
+          <label className="text-sm mb-1 mt-3 text-gray-700">Stop sequences</label>
+          <label className="text-sm text-gray-500 mb-1">
+            Enter an option, then press &apos;Tab&apos; to add
+          </label>
+          <ClassifyOptions options={stop} setOptions={setStop} placeholder="" />
```
Comment (lslunis) [inline ui/components/TracePage/PromptEditor.tsx:358]: It's confusing that <ClassifyOptions> is used here, can you name the component more generically?

**Comment by lslunis** [comment]
Left a final round of comments! Only the high priority ones are blocking.

**Pull request closed without merging**